website_canonical_url/models/website.py

```diff
@@ -1,35 +1,75 @@
 # -*- coding: utf-8 -*-
+# Copyright 2018 Simone Orsi <simone.orsi@camptocamp.com>
 # Copyright initOS GmbH 2016
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import models, api
+from odoo import models, api, fields, exceptions, _
 from odoo.http import request
-from urlparse import urlparse, urlunparse
+from urlparse import urlparse, urljoin
 
 
 class Website(models.Model):
     _inherit = 'website'
 
+    canonical_domain = fields.Char(
+        help='Canonical domain is used to build unique canonical URLs '
+             'to make SEO happy.'
+    )
+
+    @api.constrains('canonical_domain')
+    def _check_canonical_domain(self):
+        domain = self.canonical_domain
+        if domain and not urlparse(domain).scheme:
+            raise exceptions.ValidationError(_(
+                'Canonical domain must contain protocol `http(s)://`'
+            ))
+
     @api.multi
     def get_canonical_url(self, req=None):
-        canonical_url = None
-        if req is None:  # pragma: no cover
-            req = request
-        if req and req.httprequest.path:
-            lang = self.env.lang
-            if lang == request.website.default_lang_code:
-                canonical_url = req.httprequest.path
-            else:
-                url_parts = urlparse(req.httprequest.path)
-                url_parts = list(url_parts)
-                # change the path of the url
-                url_parts[2] = lang + url_parts[2]
-                canonical_url = urlunparse(url_parts)
+        return urljoin(
+            self._get_canonical_domain(),
+            self._get_canonical_relative_url(req=req)
+        )
+
+    @api.multi
+    def _get_canonical_domain(self):
+        self.ensure_one()
+        if self.canonical_domain:
+            return self.canonical_domain
+        params = self.env['ir.config_parameter'].sudo()
+        return params.get_param('web.base.url')
+
+    @api.multi
+    def _get_canonical_relative_url(self, req=None):
+        req = req or request
+        parsed = urlparse(req.httprequest.path)
+        canonical_url = parsed.path
+        lang = (
+            getattr(req, 'lang', None) or
+            self.env.lang or
+            req.website.default_lang_code
+        )
+        lang_path = ''
+        if lang != req.website.default_lang_code:
+            # eg: language is `it_IT` but main lang is `en_US`
+            lang_path = '/%s/' % lang
```

Comment:
OK, I'm remembering the _why_ behind previous implementation. This is gonna simplify all the algorithm.

The main point is that **the request you will use will never start by the lang code**. If you read [here](https://github.com/odoo/odoo/blob/79bbd1dcb8c24d4be26807494d8d670faff9d7e3/addons/website/models/ir_http.py#L198-L202), you'll see that it gets rerouted, so your code must never care about the path starting or not by the lang, and you can just use `request.lang`.

So, this would be mostly the correct algorithm:

```python
canonical_url = req.httprequest.path
if canonical_url == "/page/homepage":
    canonical_url = "/"
if request.lang != request.website.default_lang_code:
    canonical_url = u"/{}{}".format(request.lang, canonical_url)
return canonical_url
```

Reply:
@Yajo actually is the other way around: it's rerouted [only if default lang is used](https://github.com/odoo/odoo/blob/79bbd1dcb8c24d4be26807494d8d670faff9d7e3/addons/website/models/ir_http.py#L181):
> if lang in url but not the displayed or default language --> change or remove
> or no lang in url, and lang to dispay not the default language --> add lang

so you definitely get the request w/ the lang in the path. And, this is ok and we don't have to check if the lang is there or not because is going to be there always, when needed.

The only exception to handle is the root menu item but we cannot handle /page/homepage only.
Anyway, I got the point so I'm going to adapt it.